### PR TITLE
[bug] Binomial distribution BTRS algorithm has small chance of returning -1

### PR DESCRIPTION
**Pull request by jamestwebber**
I was so excited to take advantage of #36858 getting merged that I installed the nightly build, and I'm glad I did!

It turns out that there's a _very small_ chance that the current algorithm will return a negative value (I imagine only -1 is possible but not sure about that).

Basically the logic [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Distributions.h#L198-L213), which returns a value that passes certain checks before checking if its negative. I can't figure out the particular range that causes this but could reproduce it by taking a billion samples with `count` 1 and `prob` 0.9:

```python
(
    torch.distributions.Binomial(
        total_count=torch.tensor(1.0).cuda(), probs=torch.tensor(0.9).cuda()
    ).sample(torch.Size((1000000000,))) >= 0
).all()
```

Reliably evaluates to `tensor(False, device='cuda:0')` on my machine. 100M samples usually does it but not always, so that's around the rate at which this crops up (it took me most of a whole day to run into it!). Seems to be CUDA specific, I imagine due to some esoteric reason I cannot begin to guess.

This PR tries to solve it in the most obvious way: reject negative values _before_ testing the bounding box, not after. But a better solution is probably to figure out why this occurs at all, and stop it.

### CODE REVIEW

**Comment by jerryzh168** [comment]
cc @crcrpar could you review

**Comment by crcrpar** [comment]
I didn't touch this part the last time so I believe @jamestwebber and the change itself looks good.

**Comment by jamestwebber** [comment]
I appreciate the vote of confidence, although I'm still confused about how this can happen at all.

**Comment by jerryzh168** [comment]
> I appreciate the vote of confidence, although I'm still confused about how this can happen at all.

sorry, what do you mean?

Could you request review from people? or do you just need a stamp?

**Comment by jamestwebber** [comment]
> sorry, what do you mean?

I meant that I was unable to reproduce the bug when I went through the algorithm by hand--there's some very rare condition that causes this but I didn't see how to back calculate it. It's possible I'm just failing at some basic algebra there but it's eluded me.

> Could you request review from people? or do you just need a stamp?

I don't think I know who to request. In terms of this specific PR I think the change is fine and @crcrpar 's review is sufficient. If you want to review why it's necessary I'd recommend someone who has debugged weird cuda precision problems before.

**Review by jerryzh168 — APPROVED** [review]

**Comment by jerryzh168** [comment]
OK, I will land this then

**Comment by jamestwebber** [comment]
I just tested this with pytorch-nightly and it's _still_ returning negative numbers somehow. So the source of this bug must be elsewhere in the code. This PR could potentially be reverted although I think it's harmless.
